lib/aiken/trees/mt.ak

```diff
@@ -0,0 +1,512 @@
+use aiken/bytearray.{from_string}
+use aiken/hash.{Hash, Sha2_256, sha2_256}
+use aiken/list
+use aiken/option.{choice, is_none}
+
+// Construction of the merkle tree shouldn't be done by hand, but via
+// 'from_list'.
+/// A purely functional implementation of MerkleTrees that is suitable for
+/// usage on-chain. Note, however, that the construction of 'MerkleTree' and
+/// membership proofs are still expected to happen *off-chain* while only the
+/// proof verification should be done on-chain.
+///
+/// Code ported to Aiken from Hydra (https://github.com/input-output-hk/hydra/blob/master/plutus-merkle-tree/src/Plutus/MerkleTree.hs)
+/// A MerkleTree representation, suitable for on-chain manipulation.
+pub type MerkleTree {
+  Empty
+  Leaf { value: String, hash: Hash<Sha2_256, ByteArray> }
```

Comment:
🤔 Is the `String` temporary?

Reply:
Yes, last time I tried generics everywhere it all failed.

In fact this code doesn't even compile anymore with latest aiken. I am using outdated version on another machine.

Reply:
See: https://github.com/aiken-lang/aiken/issues/367

Reply:
I can confirm that latest version of Aiken works fine with this MT.

Reply:
Unfortunately generics based implementation doesn't work: https://github.com/aiken-lang/aiken/issues/376

Reply:
Turned String -> ByteArray so now MerkleTree can back up anything.